lib/std/os.zig

```diff
@@ -6260,32 +6260,64 @@ pub const CopyFileRangeError = error{
-/// characteristics than transferring data from kernel to user space and back, such as with
-/// `pread` and `pwrite` calls.
-///
-/// `fd_in` must be a file descriptor opened for reading, and `fd_out` must be a file descriptor
-/// opened for writing. They may be any kind of file descriptor; however, if `fd_in` is not a regular
-/// file system file, it may cause this function to fall back to calling `pread` and `pwrite`, in which case
-/// atomicity guarantees no longer apply.
+/// `fd_in` must be a file descriptor opened for reading, and `fd_out` must be
+/// a file descriptor opened for writing. They may be any kind of file
+/// descriptor; however, if `fd_in` is not a regular file system file, it may
+/// cause this function to fall back to calling `pread` and `pwrite`, in which
+/// case atomicity guarantees no longer apply.
 ///
-/// If `fd_in` and `fd_out` are the same, source and target ranges must not overlap.
-/// The file descriptor seek positions are ignored and not updated.
-/// When `off_in` is past the end of the input file, it successfully reads 0 bytes.
+/// If `fd_in` and `fd_out` are the same, source and target ranges must not
+/// overlap. The file descriptor seek positions are ignored and not updated.
+/// When `off_in` is past the end of the input file, it successfully reads 0
+/// bytes.
 ///
-/// `flags` has different meanings per operating system; refer to the respective man pages.
-///
-/// These systems support in-kernel data copying:
-/// * Linux 4.5 (cross-filesystem 5.3)
+/// Depending on the system, a few mechanisms are tried:
 ///
-/// Other systems fall back to calling `pread` / `pwrite`.
+/// * Linux 4.5+: ioctl.FICLONERANGE: atomic, O(1), the fastest method. Uses
+///   copy-on-write, therefore saves disk space and time. As of Linux 5.19,
+///   available on btrfs, cifs, nfs, ocfs2, overlayfs and xfs. The source and
+///   destination must be on the same file system.
+/// * Linux 4.5+: `copy_file_range(2)` via a libc wrapper (if libc is linked)
+///   or a syscall. This works at the block layer, so cross-filesystem
+///   in-kernel copying (Linux 5.3+) is possible.
+/// * Everything else: `pread`/`pwrite`.
 ///
 /// Maximum offsets on Linux are `math.maxInt(i64)`.
-pub fn copy_file_range(fd_in: fd_t, off_in: u64, fd_out: fd_t, off_out: u64, len: usize, flags: u32) CopyFileRangeError!usize {
+pub fn copy_file_range(fd_in: fd_t, off_in: u64, fd_out: fd_t, off_out: u64, len: usize) CopyFileRangeError!usize {
+    const ficlone_range = comptime builtin.os.isAtLeast(.linux, .{ .major = 4, .minor = 5 }) orelse true;
+
+    if (ficlone_range) {
+        const arg = linux.FICLONERANGE_arg{
+            .src_fd = fd_in,
+            .src_offset = off_in,
+            .src_length = len,
+            .dest_offset = off_out,
+        };
+        while (true) {
+            const rc = system.ioctl(fd_out, linux.T.FICLONERANGE, @ptrToInt(&arg));
```

Comment:
Did you benchmark the impact when used on filesystems which do not support CoW clones? In that case, this adds an extra system call for each run

As of at least Ubuntu 20.04, the default filesystem is ext4 which does not support CoW clones

Reply:
I should have put this to the commit message; I am not overly concerned with the extra syscall, since coreutils is doing reflinking by default for `cp`; so I punted on the costs.

I see some other issues with the PR, marking as draft.

Reply:
Linux kernel refuses to accept the argument, returns EINVAL for anything I give it. I will try with a C version, but not today.

Reply:
Thanks for pointing out the number of syscalls. We may get rid of one in https://github.com/ziglang/zig/pull/12491